### ISSUE TITLE
keras.Model.fit cannot handle variable epoch sizes

### ISSUE THREAD

**Issue by jackd** [tensorflow/tensorflow#41571]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): ubuntu 18.04
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): v2.2.0-0-g2b96f3662b 2.2.0
- Python version: 3.6.9

**Describe the current behavior**
`keras.Model.fit` with finite-length dataset inputs can run without `steps_per_epoch`. However, if the number of batches in the epoch changes between epochs, this can lead to:
* truncation of epoch if subsequence epochs are longer than the initial epoch; and
* truncation of training if subsequence epochsa re shorter than the initial epoch.

While not a common occurance, this happens e.g. in graph neural network applications where it makes more sense to batch according to some maximum number of edges/nodes in a graph as opposed to a fixed batch size. A fixed number of graphs in a dataset can result in a variable number of batches when these are shuffled.

**Describe the expected behavior**
Each epoch of `fit` should iterate over a dataset in it's entirety when `steps_per_epoch` is not provided, and run to the defined number of epochs (unless terminated prior with callbacks), even if the number of steps varies between epochs. Alternatively, a special `steps_per_epoch` value (say, -1) should be able to be provided to specify a dynamic epoch size.
**Standalone code to reproduce the issue**
[Colab](https://colab.research.google.com/drive/1HWPtXa2N_EcSFjhcFDquxBrl8RlZvB9y?usp=sharing)

**Other info / logs**
From colab above, when subsequence epochs are shorter than the first:
```bash
WARNING:tensorflow:Your input ran out of data; interrupting training. Make sure that your dataset or generator can generate at least `steps_per_epoch * epochs` batches (in this case, 30 batches). You may need to use the repeat() function when building your dataset.
```

**Comment by amahendrakar** [comment]
Was able to reproduce the issue with TF v2.2 and TF-nightly. Please find the gist of it [here](https://colab.research.google.com/gist/amahendrakar/81bba42e7be6d5f5b31523988f902207/41571-tf-nightly.ipynb). Thanks!

**Comment by omalleyt12** [comment]
@jackd Thanks for the issue! I'd recommend using a Dataset that contains every batch you want to train on, and setting steps_per_epoch as an arbitrary number, so that epoch boundaries are just a way to check on the progress of training in your case. If that doesn't work, then you can always write a custom training loop for your use case: https://keras.io/guides/writing_a_training_loop_from_scratch/
